avae/vis.py

```diff
@@ -1145,6 +1145,122 @@ def recon_plot(img, rec, label, data_dim, mode="trn", epoch=0, writer=None):
+    x, xs, ys, vae, device, data_dim, plots_config, poses=None
+):
+    """Visualise the interpolation of latent space between 4 randomly selected encodings.
+    The number of plots and the number of interpolation steps is modifyable.
+
+    Parameters
+    ----------
+    x: torch.Tensor
+        A sample batch. we extract 4 random images from this
+    xs: list
+        the list of all latent vectors
+    ys: list
+        the list of all labels for each latent vector in xs
+    vae: torch.nn.Module
+        Affinity vae model.
+    device: torch.device
+        Device to run the model on.
+    data_dim: int
+        the size of spatial dimensions of the input data (2:2D, 3:3D)
+    plots_config: List
+        A list containing the number of plots to be generated and the number of interpolation steps.
+    poses: list
+        List of pose vectors.
+    """
+    logging.info(
+        "################################################################",
+    )
+    logging.info(
+        "Visualising Latent Interpolation between 4 randomly selected encodings ...\n"
+    )
+    padding = 0
+    enc = []
+    decoded_images = []
+    dsize = x[0].shape[-data_dim:]
+    classes = np.unique(np.asarray(ys))
+    latent_dim = xs.shape[1]
+
+    # Number of plots (each have 4 random corners)
+    plots_config = plots_config.replace(" ", "").split(",")
+
+    # Number of interpolation steps
+    num_steps = int(plots_config[1])
+
+    if poses is not None:
+        pose_mean = np.mean(poses)
+
+    for num_fig in range(int(plots_config[0])):
+        enc = []
+        decoded_images = []
+
+        draw_four = random.sample(range(len(classes)), k=4)
+        selected_classes = [classes[index] for index in draw_four]
+        for idx in draw_four:
+            lat = np.take(
+                xs,
+                random.sample(list(np.where(ys == classes[idx])[0]), k=1),
+                axis=0,
+            )
+            enc.append(lat)
+
+        enc = np.asarray(enc)
+        alpha_values = torch.linspace(0, 1, num_steps)
+        beta_values = torch.linspace(0, 1, num_steps)
+        interpolation_grid = torch.zeros((num_steps, num_steps, latent_dim))
+
+        for i, h in enumerate(alpha_values):
+            for j, v in enumerate(beta_values):
+
+                # bilinear interpolation in the latent space
+                interpolated_z = (
+                    (1 - h) * (1 - v) * enc[0]
+                    + h * (1 - v) * enc[1]
+                    + (1 - h) * v * enc[2]
+                    + h * v * enc[3]
+                )
+                interpolation_grid[i, j, :] = interpolated_z
+
+        # Decode the interpolated encoding to generate an image
+        with torch.no_grad():
+            decoded_images = vae.decoder(
+                interpolation_grid.view(-1, latent_dim).to(device=device),
+                (
+                    torch.Tensor(num_steps * num_steps, poses[0].shape[0])
+                    + pose_mean
```

Comment:
This line can make the pose randomly incorrect therefore plotting misleading results.

Empty uninitialised tensors can carry some values left over in memory so it is bad practice to add values to empty uninitialised data structures. Instead of `torch.Tensor` this should be `torch.zeros` which initialises all reserved memory to a value 0.

When `torch.Tensor` is used, occasionally the plot will contain empty boxes. They are not actually empty but due to dynamic range of the `imshow` plots they appear to be. In reality the mean pose value was added to a very large value already contained in the tensor resuliting in a very large pose value for that image - which resulted in much larger pixel values when decoded.

Here's an extreme example of this where plot was all empty but for one image that contained a value of 5194705394675261264393340065310310400 for the pose.
![Screenshot from 2023-11-27 14-38-29](https://github.com/alan-turing-institute/affinity-vae/assets/6930946/dbebd17d-ce14-4806-be93-56a0ff19201b)

`tensor([[5194705394675261264393340065310310400.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [-214192887127601053696.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.5039],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000],
        [     0.0000]])
`

Reply:
Ok, this maybe why I wasn't getting this on my computer but it happens on linux ! thanks a lot for spotting that , it was baffling me !